### PR TITLE
Release checklist

### PR DESCRIPTION
**Pull request by mojoaxel**
This replaces the existing [release docs](https://github.com/almende/vis/blob/v4.20.1-SNAPSHOT/misc/how_to_publish.md) with a new checklist for releases.

I hope this helps maintainers to release more often.

### CODE REVIEW

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Release Checklist
+
+## Communication
+- [ ] Create a new issue and copy&past this checklist into it (Yeah! First Step done!)
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: past -> paste

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Talk to the team: Who should make the release?
+- [ ] Announce a "Code-Freeze". No new Pull-Request until the release is done!
+- [ ] Checkout if we have MAJOR or MINOR changes. If not we do a PATCH release.
+- [ ] The new version will be: `vX.X.Y`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: Strictly speaking it should be `vX.Y.Z` (so major and minor versions can be different)

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Identify open BUGS and add them to the next PATCH milestone (optional).
+- [ ] Identify MINOR issues and add them to the next MINOR milestone (optional).
+
+## Update to the newest version
+- [ ] Update to the current version: `git checkout develop && git pull`.
+- [ ] Create a new release branch. (`git checkout -b vX.X.Y develop`)
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+## Build & Test
+- [ ] Update the version number of the library in `package.json` (remove the "SNAPSHOT").
+- [ ] Build the library: `npm prune && rm -rf node_modules && npm install && npm run build && npm run test`
+- [ ] Open some of the examples in your browser and visually check if it works as expected! (*We need automated tests for this!*)
+
+## History
+(*THIS IS A LOT OF WORK! WE SHOULD TRY TO automate this in the future!!*)
+
+- [ ] Get all commits since the last release: ```git log `git describe --tags --abbrev=0`..HEAD --oneline > .commits.tmp```
+- [ ] Open ".commity.tmp". and remove all commit before the last release.
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `.commits.tmp` (also "all commit" -> "all commits"

Seems like there should be a way to get git to do this part, optimisation for later.

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Open every commit in GitHub and move every issue/pull-request to the current milestone.
+- [ ] Transfer all Commit-Messages/issues to "HISTORY.md" starting at the button.
+ - Keep the order of the commits. Older commits are lower newers are higher.
+ - Bug-Fixes start with `FIX #issue:`
+ - New Features start with `FEAT #issue:`
+
+## Commit
+- [ ] Commit the new version: `git commit -am "Release vX.X.Y"`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Wait until somebody of the team looked over your changes and merges the Pull-Request.
+
+### Update Master
+We don't merge the development branch to the master because the master branch is different to the develop-Branch. The master branch has a dist and test folder and does not generate Source-Maps.
+
+If we would merge the development branch would overwrite this. To solve this we use rebase instead:
+
+- [ ] Update: `git fetch && git checkout develop && git pull`
+- [ ] Rebase the `master` branch on the `develop` branch: `git checkout master && git rebase develop`
+- [ ] Generate new dist files: `npm prune && rm -rf node_modules && npm install && npm run build && npm run test && git commit -am "generated dist files for vX.X.Y"
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Create a version tag: `git tag "vX.X.Y"`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] [Remove the protection](https://github.com/almende/vis/settings/branches/master) from `master`.
+- [ ] FORCE-Push the branches to github: `git push --force && git push --tag`
+- [ ] [Re-Enable branch protection](https://github.com/almende/vis/settings/branches/master) (enable ALL checkboxes) for `master`.
+- [ ] Publish with npm: `npm publish` (check [npmjs.com](https://www.npmjs.com/package/vis))
+- [ ] Create a [new Release](https://github.com/almende/vis/releases/new) with the tang and the name "vX.X.Y" and copy the data vom [HISTORY.md](../HISTORY.md) into the body.
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: "tang" -> "tag"?
"vom" -> "from"

`vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+
+## Test
+- [ ] Go to a temp directory (e.g. "vis_vX.X.Y"): `cd .. && mkdir vis_vX.X.Y && cd vis_vX.X.Y`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Install the library from npm: `npm init -f && npm install vis`
+- [ ] Verify if it installs the just released version, and verify if it works: `cd node_modules/vis/
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: missing ` character at the end.

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Install the library via bower: `cd ../.. && bower install vis`
+- [ ] Verify if it installs the just released version, and verify if it works: `cd bower_components/vis/`
+- [ ] Clone the master from github: `cd ../.. && git clone git@github.com:almende/vis.git`.
+- [ ] Verify if it installs the just released version, and verify if it works. `cd vis`
+
+## Update website
+- [ ] update the gh-pages branch: `git checkout gh-pages && git pull && git checkout -b "gh-pages_vX.X.Y"`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Copy the `dist` folder from the `master` branch to the `github-pages` branch in another directory, overwriting existing files: `cp -rf ../vis_vX.X.Y/vis/dist .`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Copy the `docs` folder from the `master` branch to the `github-pages` branch in another directory, overwriting existing files: `cp -rf ../vis_vX.X.Y/vis/docs .`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Copy the `examples` folder from the `master` branch to the `github-pages` branch in another directory, overwriting existing files: `cp -rf ../vis_vX.X.Y/vis/examples .`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Check if there are new or updated examples, and update the gallery screenshots accordingly.
+- [ ] Update the library version number in the `index.html` page.
+- [ ] Update the CDN links at the download section of index.html AND the CDN link at the top. (search-replace all!!)
+- [ ] Commit the changes: `git add -A && git commit -m "updates for vX.X.Y"`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Push the changes `git push --set-upstream origin gh-pages_vX.X.Y`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+## Prepare next version
+- [ ] Switch to the "develop" branch: `git checkout develop`.
+- [ ] Change version numbers in "package.json" to a snapshot version `X.X.Z-SNAPSHOT`.
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z+1`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Commit and push: `git commit -am "changed version to vX.X.Z-SNAPSHOT"`
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z+1`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+- [ ] Create new tag: `git tag vX.X.Z-SNAPSHOT`.
```
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: `vX.Y.Z+1`

File: misc/RELEASE_CHECKLIST_TEMPLATE.md (hunk 17 above)
Comment (yotamberk) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: No... This should stay `SNAPSHOT` until next version is updated

File: misc/RELEASE_CHECKLIST_TEMPLATE.md (hunk 18 above)
Comment (yotamberk) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: No... This should stay `SNAPSHOT` until next version is updated

File: misc/RELEASE_CHECKLIST_TEMPLATE.md (hunk 19 above)
Comment (yotamberk) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: No... This should stay `SNAPSHOT` until next version is updated

**Comment by yotamberk** [comment]
@bradh can you please approve this so I can merge the changes?

File: misc/RELEASE_CHECKLIST_TEMPLATE.md (hunk 19 above)
Comment (bradh) [inline misc/RELEASE_CHECKLIST_TEMPLATE.md]: @yotamberk Approved the changes. I meant that the version after vX.Y.Z should be vX.Y.Z+1-SNAPSHOT,

**Comment by mojoaxel** [comment]
@bradh Sorry I simply forgot about this one!
@yotamberk Thanks for fixing my mistakes!!

**Pull request merged**